File: ShadowFrameworkLite20/src/shadow/pipeline/SFPipeline.java
package shadow.pipeline;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;


/**
 * The Shadow Framework Pipeline is a Virtual Rendering Pipeline which the
 * ShadowFramework will implement through some OpenGL version.
 * 
 * The ShadowFramework Pipeline allows a combination of five ShaderComponents,
 * one for each programmable Virtual Shadow Framework Rendering steps:
 * Tessellation, Primitive, Transform, Material and Lighting.
 * 
 * @author Alessandro Martinelli
 */
public class SFPipeline {
	
	private static class SFProgramTrace implements Comparable<SFProgramTrace>{
		
		private String[] modulesNames;
		private SFProgram program;
		
		private SFProgramTrace(String[] modulesNames) {
			super();
			this.modulesNames = modulesNames;
		}

		@Override
		public int compareTo(SFProgramTrace o) {
			if(modulesNames.length!=o.modulesNames.length)
				return (modulesNames.length-o.modulesNames.length);
			for (int i = 0; i < modulesNames.length; i++) {
				int compare=modulesNames[i].compareTo(o.modulesNames[i]);
				if(compare!=0)
					return compare;
			}
			
			return 0;
		}
	}
	
	
	// Pipeline components.
	private HashMap<String, SFProgramComponent> components=new HashMap<String, SFProgramComponent>();
	private HashMap<String, SFProgramModule> modules=new HashMap<String, SFProgramModule>();
	private HashMap<String, SFPipelineStructure> structures=new HashMap<String, SFPipelineStructure>();
	private HashMap<String, SFPrimitive> primitives=new HashMap<String, SFPrimitive>();
	
	private List<SFProgramTrace> traces=new ArrayList<SFProgramTrace>();

	private SFProgramBuilder sfProgramBuilder;
	private SFPipelineGraphics sfPipelineGraphics;
	private SFPipelineMemory sfPipelineMemory;
	private SFTexturePipeline sfTexturePipeline;

	private static SFPipeline pipeline=new SFPipeline();
	
	private SFPipeline() {
		super();
	}

	public static SFPipeline getPipeline() {
		return pipeline;
	}

	public static void setSfTexturePipeline(SFTexturePipeline sfTexturePipeline) {
		pipeline.sfTexturePipeline=sfTexturePipeline;
	}

	public static SFProgramBuilder getSfProgramBuilder() {
		return pipeline.sfProgramBuilder;
	}

	public static void setSfProgramBuilder(SFProgramBuilder sfProgramBuilder) {
		pipeline.sfProgramBuilder=sfProgramBuilder;
	}
	
	public static SFTexturePipeline getSfTexturePipeline() {
		return pipeline.sfTexturePipeline;
	}
	
	public static SFPipelineMemory getSfPipelineMemory() {
		return pipeline.sfPipelineMemory;
	}

	public static void setSfPipelineMemory(SFPipelineMemory sfPipelineMemory) {
		pipeline.sfPipelineMemory = sfPipelineMemory;
	}

	public static SFPipelineGraphics getSfPipelineGraphics() {
		return pipeline.sfPipelineGraphics;
	}

	public static void setSfPipelineGraphics(
			SFPipelineGraphics sfPipelineGraphics) {
		pipeline.sfPipelineGraphics=sfPipelineGraphics;
	}

	// Pipeline components.
	public static void loadShaderComponent(String code,
			SFProgramComponent component) {
		pipeline.components.put(code,component);
	}
	
	// Pipeline components.
	public static void loadPrimitive(String code,
			SFPrimitive component) {
		pipeline.primitives.put(code,component);
	}

	// Pipeline components.
	public static void loadShaderModule(String code,
			SFProgramModule component) {
		pipeline.modules.put(code,component);
	}


	public static SFPipelineElement getModule(String structureCode) {
		SFPipelineStructure structure=pipeline.structures.get(structureCode);
		if (structure != null) {
			return structure;
		}
		return pipeline.components.get(structureCode);
	}
	
	public static SFProgramModule getProgramModule(String name){
		SFProgramModule module=pipeline.modules.get(name);
		if(module==null){
			SFProgramComponent component=(SFProgramComponent)(getModule(name));
			if(component!=null){
				module=new SFProgramModule();
				SFProgramComponent[] components={component};
				module.setComponents(components);
			}
		}
		return module;
	}

	public static SFPrimitive getPrimitive(String name){
		SFPrimitive module=pipeline.primitives.get(name);
		return module;
	}

	public static SFPipelineStructure getStructure(String structureCode) {
		return pipeline.structures.get(structureCode);
	}

	// Pipeline components.
	public static void loadStructure(String code, SFPipelineStructure component) {
//		if(pipeline.structures.containsKey(code))
//			throw new RuntimeException("Structure "+code+" has already been loaded");
		pipeline.structures.put(code,component);
	}

	/**
	 * Generate a unique Static SFShader given its parameters, or create it if
	 * it does not exists.
	 * 
	 * @param tessellator
	 * @param primitive
	 * @param transform
	 * @param material
	 * @param light
	 * @return
	 */
	public static SFProgram getStaticProgram(SFPrimitive primitive, String transform, String material, String light) {

		SFProgramTrace trace=generateTrace(primitive.getName(), transform, material, light);
		
		int index = Collections.binarySearch(pipeline.traces, trace); 
		
		if(index>=0){
		
			return pipeline.traces.get(index).program;
			
		}else{
			SFProgram program=pipeline.sfProgramBuilder.generateNewProgram();

			program.setPrimitive(primitive);
			SFProgramModule transformP=getProgramModule(transform);
			SFProgramModule materialP=getProgramModule(material);
			SFProgramModule lightP=getProgramModule(light);
			program.setTransform(transformP);
			program.setMaterial(materialP);
			program.setLightStep(lightP);
			trace.program=program;
			
			pipeline.traces.add(trace);
			Collections.sort(pipeline.traces);
			
			// Now program is returned, and it is ready to be used
			return program;
		}
		
		
		

	}

	public static SFProgramTrace generateTrace(String primitive, String transform, String material, 
			String light) {
		
		String[] modules = {primitive,transform,material,light};
	
		SFProgramTrace trace=new SFProgramTrace(modules);
		return trace;
	}
	
	public static SFProgramTrace generateTrace(String material, 
			String light) {
		
		String[] modules = {material,light};
	
		SFProgramTrace trace=new SFProgramTrace(modules);
		return trace;
	}

	/**
	 * Generate a unique Static SFShader given its parameters, or create it if
	 * it does not exists.
	 * 
	 * @param tessellator
	 * @param primitive
	 * @param transform
	 * @param material
	 * @param light
	 * @return
	 */
	public static SFProgram getStaticImageProgram(String material, String light) {

		SFProgramTrace trace=generateTrace(material, light);
		
		int index = Collections.binarySearch(pipeline.traces, trace); 
		
		if(index>=0){
		
			return pipeline.traces.get(index).program;
			
		}else{
			SFProgram program=pipeline.sfProgramBuilder.generateImageProgram();

			program.setMaterial(getProgramModule(material));
			program.setLightStep(getProgramModule(light));
			
			trace.program=program;
			
			pipeline.traces.add(trace);
			Collections.sort(pipeline.traces);
			
			// Now program is returned, and it is ready to be used
			return program;
		}
	}
	
}

File: SFDataApplicationTestC++/shadow/pipeline/SFPipeline.cpp
#include "SFPipeline.h"



namespace sf{

SFPipeline SFPipeline::pipeline;


int SFProgramTrace_compareTo(SFProgramTrace a,SFProgramTrace b) {
    if(a.modulesNames.size()!=b.modulesNames.size())
        return (a.modulesNames.size()-b.modulesNames.size());
    for (unsigned int i = 0; i < a.modulesNames.size(); i++) {
        int compare=a.modulesNames[i].compare(b.modulesNames[i]);
        if(compare!=0)
            return compare;
    }
    
    return 0;
}


SFPipeline::SFPipeline() {
    this->sfPipelineGraphics=0;
    this->sfPipelineMemory=0;
    this->sfProgramBuilder=0;
    this->sfTexturePipeline=0;
    this->sfPipelineNets=0;
}

SFPipeline SFPipeline::getPipeline() {
    return pipeline;
}

void SFPipeline::setSfTexturePipeline(SFTexturePipeline* sfTexturePipeline) {
    pipeline.sfTexturePipeline=sfTexturePipeline;
}

SFProgramBuilder* SFPipeline::getSfProgramBuilder() {
    return pipeline.sfProgramBuilder;
}

void SFPipeline::setSfProgramBuilder(SFProgramBuilder* sfProgramBuilder) {
    pipeline.sfProgramBuilder=sfProgramBuilder;
}

SFTexturePipeline* SFPipeline::getSfTexturePipeline() {
    return pipeline.sfTexturePipeline;
}

SFPipelineMemory* SFPipeline::getSfPipelineMemory() {
    return pipeline.sfPipelineMemory;
}

void SFPipeline::setSfPipelineMemory(SFPipelineMemory* sfPipelineMemory) {
    pipeline.sfPipelineMemory = sfPipelineMemory;
}

SFPipelineGraphics* SFPipeline::getSfPipelineGraphics() {
    return pipeline.sfPipelineGraphics;
}

void SFPipeline::setSfPipelineGraphics(
                                  SFPipelineGraphics* sfPipelineGraphics) {
    pipeline.sfPipelineGraphics=sfPipelineGraphics;
}

SFPipelineNets* SFPipeline::getSfPipelineNets() {
    return pipeline.sfPipelineNets;
}

void SFPipeline::setSfPipelineNets(SFPipelineNets* sfPipelineNets) {
    pipeline.sfPipelineNets=sfPipelineNets;
}

// Pipeline components.
void SFPipeline::loadShaderComponent(string code,
                                SFProgramComponent* component) {
    pipeline.components[code]=component;
}

// Pipeline components.
void SFPipeline::loadPrimitive(string code,
                          SFPrimitive* component) {
    pipeline.primitives[code]=component;
}

// Pipeline components.
void SFPipeline::loadShaderModule(string code,
                             SFProgramModule* component) {
    pipeline.modules[code]=component;
}


SFPipelineElement* SFPipeline::getModule(string structureCode) {
    SFPipelineStructure* structure=pipeline.structures[structureCode];
    if (structure != 0) {
        return structure;
    }
    return pipeline.components[structureCode];
}

SFProgramModule* SFPipeline::getProgramModule(string name){
    SFProgramModule* module=pipeline.modules[name];
    if(module==0){
        SFProgramComponent* component=(SFProgramComponent*)(getModule(name));
        if(component!=0){
            module=new SFProgramModule();
            vector<SFProgramComponent*> components;
            components.clear();
            components.push_back(component);
            module->setComponents(components);
        }
    }
    return module;
}

SFPrimitive* SFPipeline::getPrimitive(string name){
    SFPrimitive* module=pipeline.primitives[name];
    return module;
}

SFPipelineStructure* SFPipeline::getStructure(string structureCode) {
    return pipeline.structures[structureCode];
}

// Pipeline components.
void SFPipeline::loadStructure(string code, SFPipelineStructure* component) {
    //		if(pipeline.structures.containsKey(code))
    //			throw new RuntimeException("Structure "+code+" has already been loaded");
    pipeline.structures[code]=component;
}

/**
 * Generate a unique SFShader given its parameters, or create it if
 * it does not exists.
 *
 * @param tessellator
 * @param primitive
 * @param transform
 * @param material
 * @param light
 * @return
 */
SFProgram* SFPipeline::getStaticProgram(SFPrimitive* primitive, string transform, string material, string light) {
    
    SFProgramTrace* trace=generateTrace(primitive->getName(), transform, material, light);
    
    int index = 0;
    //TO be OPTIMIZED
    //like Collections.binarySearch(pipeline.traces, trace);
    for(unsigned int i=0;i<pipeline.traces.size();i++){
        if(SFProgramTrace_compareTo(*pipeline.traces.at(i),*trace)==0){
            index=i;
            i=pipeline.traces.size();
        }
    }
    
    if(index>=0){
        return pipeline.traces.at(index)->program;
        
    }else{
        SFProgram* program=pipeline.sfProgramBuilder->generateNewProgram();
        
        program->setPrimitive(primitive);
        SFProgramModule* transformP=getProgramModule(transform);
        SFProgramModule* materialP=getProgramModule(material);
        SFProgramModule* lightP=getProgramModule(light);
        program->setTransform(transformP);
        program->setMaterial(materialP);
        program->setLightStep(lightP);
        trace->program=program;
        
        pipeline.traces.push_back(trace);
        //TO BE PTIMIZED
        //Collections.sort(pipeline.traces);
        
        // Now program is returned, and it is ready to be used
        return program;
    }
    
    
    
    
}

SFProgramTrace* SFPipeline::generateTrace(string primitive, string transform, string material,
                                     string light) {
    
    vector<string> modules;
    modules.clear();
    modules.push_back(primitive);
    modules.push_back(transform);
    modules.push_back(material);
    modules.push_back(light);
    
    SFProgramTrace* trace=new SFProgramTrace(modules);
    return trace;
}

SFProgramTrace* SFPipeline::generateTrace(string material,
                                     string light) {
    
    
    vector<string> modules;
    modules.clear();
    modules.push_back(material);
    modules.push_back(light);
    
    SFProgramTrace* trace=new SFProgramTrace(modules);
    return trace;
}

/**
 * Generate a unique SFShader given its parameters, or create it if
 * it does not exists.
 *
 * @param tessellator
 * @param primitive
 * @param transform
 * @param material
 * @param light
 * @return
 */
SFProgram* SFPipeline::getStaticImageProgram(string material, string light) {
    
    SFProgramTrace* trace=generateTrace(material, light);
    
    int index = 0;
    //TO be OPTIMIZED
    //like Collections.binarySearch(pipeline.traces, trace);
    for(unsigned int i=0;i<pipeline.traces.size();i++){
        if(SFProgramTrace_compareTo(*pipeline.traces.at(i),*trace)==0){
            index=i;
            i=pipeline.traces.size();
        }
    }
    
    if(index>=0){
        
        return pipeline.traces.at(index)->program;
        
    }else{
        SFProgram* program=pipeline.sfProgramBuilder->generateImageProgram();
        
        program->setMaterial(getProgramModule(material));
        program->setLightStep(getProgramModule(light));
        
        trace->program=program;
        
        pipeline.traces.push_back(trace);
        //Collections.sort(pipeline.traces);
        
        // Now program is returned, and it is ready to be used
        return program;
    }
}

}
